com.unity.mobile.notifications/Runtime/iOS/Plugins/UnityNotificationData.m

```diff
@@ -0,0 +1,159 @@
+        notificationData->categoryIdentifier = (char*)[content.categoryIdentifier  UTF8String];
+
+    if (content.threadIdentifier != nil && content.threadIdentifier.length > 0)
+        notificationData->threadIdentifier = (char*)[content.threadIdentifier  UTF8String];
+
+    if ([request.trigger isKindOfClass: [UNTimeIntervalNotificationTrigger class]])
+    {
+        notificationData->triggerType = TIME_TRIGGER;
+
+        UNTimeIntervalNotificationTrigger* timeTrigger = (UNTimeIntervalNotificationTrigger*)request.trigger;
+        notificationData->timeTriggerInterval = timeTrigger.timeInterval;
+        notificationData->repeats = timeTrigger.repeats;
+    }
+    else if ([request.trigger isKindOfClass: [UNCalendarNotificationTrigger class]])
+    {
+        notificationData->triggerType = CALENDAR_TRIGGER;
+
+        UNCalendarNotificationTrigger* calendarTrigger = (UNCalendarNotificationTrigger*)request.trigger;
+        NSDateComponents* date = calendarTrigger.dateComponents;
+
+        notificationData->calendarTriggerYear = (int)date.year;
+        notificationData->calendarTriggerMonth = (int)date.month;
+        notificationData->calendarTriggerDay = (int)date.day;
+        notificationData->calendarTriggerHour = (int)date.hour;
+        notificationData->calendarTriggerMinute = (int)date.minute;
+        notificationData->calendarTriggerSecond = (int)date.second;
+    }
+    else if ([request.trigger isKindOfClass: [UNLocationNotificationTrigger class]])
+    {
+#if defined(UNITY_USES_LOCATION) && UNITY_USES_LOCATION
+        notificationData->triggerType = LOCATION_TRIGGER;
+
+        UNLocationNotificationTrigger* locationTrigger = (UNLocationNotificationTrigger*)request.trigger;
+        CLCircularRegion *region = (CLCircularRegion*)locationTrigger.region;
+
+        notificationData->locationTriggerCenterX = region.center.latitude;
+        notificationData->locationTriggerCenterY = region.center.longitude;
+        notificationData->locationTriggerRadius = region.radius;
+        notificationData->locationTriggerNotifyOnExit = region.notifyOnEntry;
+        notificationData->locationTriggerNotifyOnEntry = region.notifyOnExit;
+#endif
+    }
+    else if ([request.trigger isKindOfClass: [UNPushNotificationTrigger class]])
+    {
+        notificationData->triggerType = PUSH_TRIGGER;
+    }
+
+    if ([NSJSONSerialization isValidJSONObject: [request.content.userInfo objectForKey: @"data"]])
+    {
+        NSError *error;
+        NSData *data = [NSJSONSerialization dataWithJSONObject: [request.content.userInfo objectForKey: @"data"]
+                        options: NSJSONWritingPrettyPrinted
+                        error: &error];
+        if (!data)
+        {
+            NSLog(@"Failed parsing notification userInfo[\"data\"]: %@", error);
+        }
+        else
+        {
+            notificationData->data = (char*)[[[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding] UTF8String];
```

Comment:
Not quite sure what the fastest way is... Any suggestions?

I saw some people recommended `NSString* newStr = [NSString stringWithUTF8String:[theData bytes]];`

Reply:
I mean the data is already an UTF-8 string. What this code does is allocate a buffer to store it using NSString and conversion to UTF-16 as interim. What does notificationData->data have to point to and what is the lifetime of that thing?

Reply:
The whole process is:
1. When you schedule a notification, you have to create the `iOSNotification` which can contain customized data as `data` at managed side, then we marshall it and passed the struct to native;
2. We put the `data` into a `NSDictionary` and set the dictionary to the `UNNotificationRequest` for scheduling a notification;
3. Once the notification is delivered, a callback will be called with a UNNotification which contains a `UNNotificationRequest`, according to the document https://developer.apple.com/documentation/usernotifications/unnotification/1649324-request?language=objc, this request is an copy of the one we configure in step 2.

Actually in step 2, we simply did
```
NSDictionary* userInfo = @{
        @"showInForeground": @(data->showInForeground),
        @"showInForegroundPresentationOptions": [NSNumber numberWithInteger: data->showInForegroundPresentationOptions],
        @"data": @(data->data ? data->data : ""),
    };
```
Why do we need to do
```
    if ([NSJSONSerialization isValidJSONObject: [request.content.userInfo objectForKey: @"data"]])
    {
        NSError *error;
        NSData *data = [NSJSONSerialization dataWithJSONObject: [request.content.userInfo objectForKey: @"data"]
                        options: NSJSONWritingPrettyPrinted
                        error: &error];
        if (!data)
        {
            NSLog(@"Failed parsing notification userInfo[\"data\"]: %@", error);
        }
        else
        {
            notificationData->data = (char*)[[[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding] UTF8String];
        }
    }
```
Can we simply return the object of key @`data` as utf-8 string?

Reply:
I think we should sort out the lifetime of notificationData->data first. Now it points to temporary allocation of C string. malloc and memcpy would be safe and more efficient, except that somewhere free() would be required.

Reply:
Agreed. I mean if we can point `notificationData->data` to the data on the request which is alive longer than notificationData, rather than the temporary string, then we will be good.

Reply:
Looked into the setting `notification->data` code, so for local notification, it can only be a string. But for remote notifications, users can set a json object as customized data, this is why we do that conversion.

Reply:
Fixed at https://github.com/Unity-Technologies/com.unity.mobile.notifications/pull/46/commits/f168e836e4843175c90724834bcdb49b4684d4ed.